(DOCSP-7083): Glossary for Realm (#142)
mongodb/docs-realm: source/get-started/glossary.txt

========
Glossary
========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

.. glossary::
   :sorted:

   MongoDB Atlas

>       Cloud-hosted managed instances of MongoDB.

Author: Asked jon about this, now have a blessed definition!

   MongoDB Realm
      Cloud-hosted Backend-as-a-Service (BaaS) providing client-side
      functionality via the Realm SDK and an administrative interface

>       via a website.

Author: Sent this over to Drew, Christian, and Kenneth. We'll see what comes back from each of them.

   MongoDB Document
      See the MongoDB Manual's :manual:`definition of document
      </core/document/>`.

   Partition Key
      A field within a :term:`MongoDB Document` that determines which Realm the
      corresponding Realm Object belongs to.

>    Realm

Reviewer: In this instance I feel like we _should_ capitalize to preserve title case

      A collection of :term:`Realm Object`s sharing the same
      :term:`Partition Key` value.
      Client instances subscribe to individual Realms which then passively
      synchronize changes as network availability allows.

   Realm Database
      An :ref:`ACID`-compliant offline-first object database for mobile devices.
      Using the Realm SDK, you can query a local instance of the Realm
      Database and synchronize your local instance with data stored
      inside :term:`MongoDB Atlas`. Offline-first functionality ensures that
      reads and writes query the local database instance only. When Sync
      is enabled, a separate background thread handles synchronization
      of changes between the local instance and your instance of MongoDB
      Atlas.
   
   Realm Object
      Object that you can store in a :term:`Realm`, defined by a
      :term:`Realm Schema` through either any Realm SDK or the
      Realm Schema editor on the MongoDB Realm admin site. Object schemas
      can define primitive fields as well as embedded lists and relations.

>    Realm Platform

Author: I was on the fence about this (it's essentially synonymous with "MongoDB Realm") but this is the nail in the coffin -- I'm gonna kill this entry off. Good point!

      The combination of :term:`MongoDB Atlas` with a :term:`MongoDB Realm`
      sync-enabled mobile app.
      This platform abstracts away the network, synchronization, and
      conflict-resolution aspects of your application. Object data within
      the mobile :term:`Realm Database` synchronizes with collections of
      documents representing those objects in MongoDB Atlas. Realm
      membership is determined using a :term:`Partition Key`.

   GraphQL
      Open-source data query and manipulation language for APIs. Realm
      GraphQL supports reading and writing data as well as subscribing
      to data changes. An alternative to :term:`Realm Database` queries
      using the client's native query language.

>    Realm Compaction

Author: Going to preserve title case but when I refer to this elsewhere I will use lowercase.

      Allocation and deallocation of data within a :term:`Realm Database`
      can lead to memory fragmentation. Calling the ``compact()`` method
      reorganizes internal data structures to reduce the storage
      overhead of your local Realm instance in the event of
      fragmentation.

   Realm Schema
      Definition of all :term:`Realm Objects` that you can read or write

>       into your local :term:`Realm Instance`.

Author: Not sure if "remote" is really a meaningful term here, but I'm going to explicitly state "client Realms" and "MongoDB Atlas" because you make a good point -- this doesn't just apply locally! It is a little unfortunate that your schema != validation, though -- items can end up in Atlas that violate the schema and don't sync with Realm as a result. Because of this I'm going to add a Schema Validation glossary entry.

   Live Objects
      Instances of objects from your :term:`Realm Schema` that
      automatically update in value to match the value of the underlying
      object in the enclosing :term:`Read Transaction`. Note that the
      enclosing read transaction might not reference the latest

>       :term:`version <Realm Version>`.

Author: Definitely a bit of a judgment call here. I personally think it's an important enough point to bring up -- Live Objects are "Live" in a sense of reflecting the latest version of the object *within their enclosing _read transaction_*, not "Live" in a sense of always reflecting the most-recent data in the DB.

>       :term:`version <Realm Version>`.

Reviewer 2: I'm just saying: could this level of detail actually lead to more confusion? Is it possible to link out to the proper detail page instead?

   Write Transaction
      You can only write to a :term:`Realm Instance` during a write
      transaction.

   Read Transaction
      Every :term:`Realm Instance` uses an implicit "read transaction" to
      generate valid instances of :term:`Live Objects`. As a result,
      the ``realm.open()`` method begins a read transaction and the
      ``realm.close()`` method closes the read transaction. Read
      transactions always use the :term:`Realm Version` that was most
      current when the realm was opened unless the client explicitly
      advances the version. You can explicitly advance the version of a
      read transaction by starting or committing a write transaction on
      the same thread as the read transaction, or by calling the
      ``realm.refresh()`` method. Platforms with a built in
      :term:`Event Loop` also refresh UI-thread read transactions in

>       between instances of the event loop.

Reviewer 2: Hmm... we had not been framing reads as "transactions" before. I would read the existing pages on reads, transactions, and threading. If you think we can see where to reframe those pages and mention "read transaction" then we can keep this. Otherwise, I would suggest reframing this entry in the glossary to be more in line with how reads are explained in the existing sections. Thoughts?

Author: I think it's very possible to restructure that page to reflect this, but I'll also throw this toward Christian to see if he has any suggestions of how to word it better/make it clearer. I think it's a really important concept so I definitely want to keep it!

   Event Loop

>       Also known as the **main loop** or **main event loop** on some

Reviewer 2: I would inform this from the page on threading, which tried to generalize '"loop" threads'. Beyond that do we need to explain such a generic concept? Where else is this relevant to Realm outside of the context of refreshing realms?

Author: I think it's valuable in some sense because Event Loop/Looper/etc. threads have a special privileged status for Read Transactions that could potentially confuse users. However I think the Threading page covers this enough that we can just get rid of this.

Author: Moving some of the Threading page description into an "Event Loop" entry. I suspect we're going to have to link to this in multiple places, so I think it's best to consolidate a definition here. Let me know if you have any other ideas!

>       platforms. A programming paradigm in which a central loop waits

Reviewer 2: ```suggestion
      platforms. A programming design pattern in which a central loop waits
```

      for signals from a UI or data source before updating underlying

>       user-facing components.

Reviewer 2: Not necessary user-facing though. A run loop can exist entirely on a background thread.

>    Realm Instance

Reviewer 2: I think the distinction between this, `realm` AND `Realm App` could be more clear in your description - in other words, why do we need a distinction between "Realm Instance" and "realm"?

Author: "Realm Instance" -> programming interface
"Realm" -> collection of disparate objects that belong "together" in some sense
"realm" -> ???
"Realm App" -> an instance of mongodb realm? I think?

@nlarew @cbush pls halp

      An interface through which a client application can read and write

>       to a local :term:`Realm`.

Reviewer 2: ```suggestion
      to a local :term:`realm`.
```

   Realm Version
      A snapshot of data in a local :term:`Realm` at a given moment.

   Realm File
      File containing all data for a local :term:`Realm` for an instance
      of :term:`Realm Database`. Because Realm Database reads and
      writes Realm data using :wikipedia:`memory mapping

>       </Memory-mapped_file`, Realm files should not exceed the amount of

Reviewer 2: ```suggestion
      </Memory-mapped_file>`, Realm files should not exceed the amount of
```

      of changes between the local instance and your instance of MongoDB
      Atlas.
   
   Realm Object
      Object that you can store in a :term:`realm <Realm>`, defined by a
      :term:`Realm Schema` through either any Realm SDK or the
      Realm Schema editor on the MongoDB Realm admin site. Object schemas
      can define primitive fields as well as embedded lists and relations.

   GraphQL
      Open-source data query and manipulation language for APIs. Realm
      GraphQL supports reading and writing data as well as subscribing
      to data changes. An alternative to :term:`Realm Database` queries
      using the client's native query language.

   Compaction
      Allocation and deallocation of data within a :term:`Realm Database`
      can lead to memory fragmentation. Calling the ``compact()`` method
      reorganizes internal data structures to reduce the storage
      overhead of your local Realm instance in the event of
      fragmentation.

   Realm Schema
      Definition of all :term:`Realm Objects` that you can read or write
      into a client :term:`Realm Instance` or :term:`MongoDB Atlas`

>       instance.

Reviewer: "MongoDB Atlas instance" -> "linked MongoDB Atlas cluster"

   Realm Database
      An :ref:`ACID`-compliant offline-first object database for mobile devices.
      Using the Realm SDK, you can query a local instance of the Realm
      Database and synchronize your local instance with data stored
      inside :term:`MongoDB Atlas`. Offline-first design ensures that
      reads and writes query the local database instance only. When Sync
      is enabled, a separate background thread handles synchronization
      of changes between the local instance and your instance of MongoDB
      Atlas.
   
   Realm object
      Object that you can store in a :term:`Realm`, defined by a
      :term:`Realm Schema` through either any Realm SDK or the
      Realm Schema editor on the MongoDB Realm admin site. Object schemas
      can define primitive fields as well as embedded lists and relations.

   GraphQL
      Open-source data query and manipulation language for APIs. Realm
      GraphQL supports reading and writing data as well as subscribing
      to data changes. An alternative to :term:`Realm Database` queries
      using the client's native query language.

   compaction
      Allocation and deallocation of data within a :term:`Realm Database`

>       and/or exceeding the size of physical memory can lead to

Reviewer 3: This sentence reads weird to me? Could it instead be something like `Allocation of data within a Realm Database that exceeds the size of physical memory can lead to fragmentation on disk`

Author: Hm, my understanding is that (as with most memory allocation) a lot of object creation/deletion events can lead to fragmentation even when well below the size of physical memory. And exceeding the size of physical memory can cause fragmentation just through normal (not a lot of) object creation/no deletions necessary. Is my understanding incorrect? Happy to change it, just want to make sure we're giving the right warnings.

Reviewer 3: Fair point. I guess It was just the use of `and/or` that threw me off a little bit. But after re-reading it again, I think it makes sense to keep as is.

      an authentication provider, such as whenever a user logs into your
      :term:`Realm app`.

   scheduled trigger
      Activate a :term:`Realm function` based on a schedule that you
      define.
   
   linked cluster
      A :term:`MongoDB Atlas` database cluster that :term:`Realm clients
      <Realm client>` can query based on roles and filters or
      synchronize data to and from using :term:`Realm database`.

   system function
      :term:`Realm function` that runs as the :term:`system user`.
   
   system user
      A user that bypasses collection rules and schema validation for
      the :term:`linked cluster` CRUD and Aggregation APIs.
   
   custom user data
      Mapped :term:`MongoDB document` that stores arbitrary fields for
      each Realm user, such as preferred language or timezone.
   
   external dependencies
      An uploaded collection of npm modules that you can import into and
      call from any :term:`Realm function`.

   function context
      A variable available in any :term:`Realm function` through which
      you can access :term:`Realm app` components, including functions
      and :term:`MongoDB Atlas` data.

>    value

Reviewer 3: This term is incredibly broad and used throughout the stack. I'm not sure it is beneficial to pin this definition to it.

Author: I'll change this to Realm value (and also Realm secret) because this is a great point.

      Durability". See :wikipedia:`the Wikipedia entry for ACID
      </ACID>` for more information.

   Realm function
      Server-side logic for your Realm application. You can call
      functions from any :term:`Realm client`, from other functions,
      and using any :term:`Realm trigger`.

   Realm trigger
      Allow you to configure responses to :term:`data changes
      <database trigger>`, :term:`user logins <authentication trigger>`,
      and :term:`scheduled events <scheduled trigger>`. Every trigger
      activation calls a :term:`Realm function`.

   database trigger
      Activate a :term:`Realm function` whenever a document is added,
      updated, or removed in a :term:`linked cluster`.

   authentication trigger
      Activate a :term:`Realm function` whenever a user interacts with
      an authentication provider, such as whenever a user logs into your
      :term:`Realm app`.

   scheduled trigger
      Activate a :term:`Realm function` based on a schedule that you
      define.
   
   linked cluster
      A :term:`MongoDB Atlas` database cluster that :term:`Realm clients
      <Realm client>` can query based on roles and filters or

>       synchronize data to and from using :term:`Realm database`.

Author: I went back and forth on this. QueryAnywhere still exists and is used by some Stitch users today (who won't be going away) so I think it's still worth mentioning this ability. I agree that it's a little frustrating though.

>       synchronize data to and from using :term:`Realm database`.

Author: used your points above + a note about how you can access the data using realm database OR the mongodb atlas service. I think that's good for now -- as services evolve and our approach changes, we can reconsider this one!

   system function
      :term:`Realm function` that runs as the :term:`system user`.
   
   system user
      A user that bypasses collection rules and schema validation for
      the :term:`linked cluster` CRUD and Aggregation APIs.
   
   custom user data
      Mapped :term:`MongoDB document` that stores arbitrary fields for

>       each Realm user, such as preferred language or timezone.

Reviewer: - stores arbitrary fields -> contains data about
- each -> a given

Can we break into two sentences? The second one can give more contextual flavor to the data:

"For example, you may store a user's preferred language, local timezone, or an ID from an external service."

   
   external dependencies
      An uploaded collection of npm modules that you can import into and

>       call from any :term:`Realm function`.

Reviewer: One or more JavaScript modules from `npm` that you have uploaded to your Realm application. You can import and use external dependencies from any Realm function.

   function context
      A variable available in any :term:`Realm function` through which
      you can access :term:`Realm app` components, including functions

>       and :term:`MongoDB Atlas` data.

Reviewer: A _global_ variable that allows you to interface with other components of your application from a Realm function

   Realm value
      A named reference to a piece of static data stored by
      :term:`MongoDB Realm`, accessible through :term:`Realm functions`

>       and :term:`Realm rules`.

Reviewer: A static _JSON_ value or Realm Secret defined in MongoDB Realm that you can access from Realm functions and rule expressions.

   Realm rules
      True or false expressions that determine whether or not a user can

>       modify, insert, or read a given piece of data.

Reviewer: - **Boolean** expressions
- Not sure about this since rules are more than just the expressions. The [JSON Expressions](https://docs.mongodb.com/stitch/services/json-expressions/) page sorta makes it seem that way but it's also generic for service rules where they really are more or less the whole thing.

Author: reworded since this applies to functions as well

   Realm secret
      A private value stored in :term:`MongoDB Realm` that cannot be
      directly read, but can be linked in authentication providers and

>       service configurations.

Reviewer: You _can_ directly read it if you assign it to a value

   Realm UI
      A front-end from which an administrator can modify your
      :term:`MongoDB Realm` configuration with tasks including defining
      :term:`functions <Realm function>`, creating :term:`triggers
      <Realm trigger>`, linking :term:`MongoDB Atlas` clusters, and

>       more.

Reviewer: - maybe s/front-end/web interface
- Not just modify your Realm configuration, it's the default interface to create/update/manage any Realm app.
- "administrator" feels off here. The UI is for app developers and anyone else that wants to modify the app or see logs.

Author: reworded to cover all of these points, thanks!

   realm-cli
      A command line interface from which an administrator can modify
      your :term:`MongoDB Realm` configuration with tasks including
      defining :term:`functions <Realm function>`, creating
      :term:`triggers <Realm trigger>`, linking :term:`MongoDB Atlas`

>       clusters, and more.

Reviewer: - `realm-cli` is the binary name, probably better to call this **Realm CLI** for parity with **Realm UI**
- Same comments vis-a-vis "Realm configuration" and "administrator" as above

   Realm Administrative API
      An REST API through which you can perform a number of

>       administrative tasks.

Reviewer: - Realm **Administration** API
- An -> A
- nit: Technically not a REST API since it violates HATEOAS and has at least one RPC endpoint.
  - Maybe just "An HTTP API..."?
